stm32 xspi driver supports Memory mapped mode for XiP

FRASTM opened the pull request:
Like the flash_stm32_qspi or flash_stm32_ospi drivers, this PR adds the Memory Mapped mode
so that code can eXecute In Place from the external  NOR flash 
This is validated on the stm32H5 serie and demonstrated on the stm32h573i_dk disco kit (Macronix MX25LM51245 octo-NOR ).

Running the samples/application_development/code_relocation_nocopy/ with CONFIG_STM32_MEMMAP=y

```
*** Booting Zephyr OS build v3.6.0-5497-g1db55c1f8901 ***                       
Address of main function 0x80005e1                                              
Address of function_in_ext_flash 0x90000001                                     
Address of var_ext_sram_data 0x200000a0 (10)                                    
Address of function_in_sram 0x20000001                                          
Address of var_sram_data 0x200000a4 (10)                                        
Hello World! stm32h573i_dk
```                                                      

Requires https://github.com/zephyrproject-rtos/zephyr/pull/73220

FRASTM committed:
boards: arm: stm32h573 disco kit can flash on external NOR

Configure the external NOR flash memory to be executable.
Change the board.cmake to add the external loader for
flashing (version that do not mass-erase the external flash).
The address of the external flash is not needed
when flashing a hex file.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
CI failure : west build -p auto -b nrf54h20dk/nrf54h20/cpuapp samples/drivers/spi_flash due to https://github.com/zephyrproject-rtos/zephyr/issues/73005

FRASTM committed:
drivers: flash: stm32 xspi driver supports the MemoryMapped mode

Adding the support of the memoryMapped mode
to the flash_stm32_xspi driver through the CONFIG_STM32_MEMMAP
That will allow XiP from the external octo_flash
Erase and Write operation must Abort operations.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: code relocation in external memory of stm32h5 disco kit

Define the configuration to run the code_relocation
application on the external memory xspi flash
of the stm32h573i_dk disco kit in XIP

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on df788d0b41300473f047a1497fae57e8ab351f26

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -1196,7 +1389,9 @@ static int flash_stm32_xspi_write(const struct device *dev, off_t addr,
 			break;
 		}
 	}
+	goto write_end;
```
goto not required here.

erwango commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -1196,7 +1389,9 @@ static int flash_stm32_xspi_write(const struct device *dev, off_t addr,
 			break;
 		}
 	}
+	goto write_end;
```
goto not required here, but required after stm32_xspi_mem_ready.

FRASTM replied:
agree, 
moved to line 1347
```
	if (ret != 0) {
		LOG_ERR("XSPI: write not ready");
		goto write_end;
	}
```
that was just keeping the same structure as flash_stm32_ospi driver

erwango reviewed (changes requested):
Otherwise LGTM

FRASTM committed:
samples: drivers spi_flash on the external NORof stm32h5 disco kit

Define the configuration to run the test on the external memory
xspi flash of the stm32h573i_dk disco kit in MemoryMapped Mode
Reading is done with memcopy and erase/write are aborting operation
to send the appropriate command

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding the configuration to execute the samples/drivers/spi_flash/ on the stm32h573i_dk with the external NOR in **MemoryMapped** mode
